libs/shared/src/lib/survey/global-properties/reference-data.ts

```diff
@@ -245,121 +258,138 @@ export const render = (
+              isEqual(choice.id, omit(valueItem.id as any, 'pos'))
+            );
+          }
+        });
+    } else {
+      element.choices = [];
     }
-    // Prevent selected choices to be removed when sending the value
-    question.clearIncorrectValuesCallback = () => {
-      // console.log(question.visibleChoices);
-      // console.log(question.value);
-    };
+  };
+
+  const initChoices = (question: Question, element: any) => {
     // look on changes
-    question.registerFunctionOnPropertyValueChanged(
+    element.registerFunctionOnPropertyValueChanged(
       'referenceData',
       (value: string) => {
-        question.referenceDataDisplayField = undefined;
+        element.referenceDataDisplayField = undefined;
         if (!value) {
-          question.referenceDataDisplayField = undefined;
-          question.referenceDataFilterFilterFromQuestion = undefined;
-          question.referenceDataFilterForeignField = undefined;
-          question.referenceDataFilterFilterCondition = undefined;
-          question.referenceDataFilterLocalField = undefined;
+          element.referenceDataDisplayField = undefined;
+          element.referenceDataFilterFilterFromQuestion = undefined;
+          element.referenceDataFilterForeignField = undefined;
+          element.referenceDataFilterFilterCondition = undefined;
+          element.referenceDataFilterLocalField = undefined;
         }
       }
     );
-    question.registerFunctionOnPropertyValueChanged(
+    element.registerFunctionOnPropertyValueChanged(
       'isPrimitiveValue',
-      updateChoices
+      updateChoices(question, element)
     );
-    question.registerFunctionOnPropertyValueChanged(
+    element.registerFunctionOnPropertyValueChanged(
       'referenceDataDisplayField',
-      updateChoices
+      updateChoices(question, element)
     );
 
     // Look for foreign question changes if needed for filter
     const foreignQuestion = (question.survey as SurveyModel)
       .getAllQuestions()
       .find(
-        (x: any) => x.name === question.referenceDataFilterFilterFromQuestion
+        (x: any) => x.name === element.referenceDataFilterFilterFromQuestion
       ) as QuestionSelectBase | undefined;
     foreignQuestion?.registerFunctionOnPropertyValueChanged(
       'value',
-      updateChoices
+      updateChoices(question, element)
     );
+  };
+
+  if (isSelectQuestion(questionElement)) {
+    const question = questionElement as QuestionSelectBase;
+
+    if (!question.referenceDataChoicesLoaded && question.referenceData) {
+      referenceDataService
+        .cacheItems(question.referenceData)
+        .then(() => updateChoices(question, question));
+      question.referenceDataChoicesLoaded = true;
+    }
+    // Prevent selected choices to be removed when sending the value
+    question.clearIncorrectValuesCallback = () => {
+      // console.log(question.visibleChoices);
+      // console.log(question.value);
+    };
```

Comment:
@estelafs Is this needed? I didn't notice anything different by removing it?

Reply:
@matheus-relief I really don't know, is a old code and i just didn't removed it because it was already commented (like  if was to be saved but not used)